flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java

```diff
@@ -203,6 +207,11 @@ protected void handleStateChange(ConnectionState newState) {
 		}
 	}
 
+	private void onReconnectedConnectionState() {
+		// check whether we find some new leader information in ZooKeeper
+		retrieveLeaderInformationFromZooKeeper();
```

Comment:
When this is retrieved, I believe it will be equal to the last value that the node cache had before disconnect except when the NodeCache has had time to update the value. On reconnect, the NodeCache sets up an asynchronous fetch of the value so the NodeCache will usually not have had time to update. If the leader changes when the zookeeper seession is lost (which will always be the case if the leader was the one that lost connection), the value that is retrieved here will likely be wrong. It seems like notifyLossOfLeader was implemented to prevent Flink nodes from having the wrong or stale leader. Is it less of a big deal for the Flink node to have the wrong leader in this case than it is in the case that notifyLossOfLeader is trying to solve?

Reply:
I think you are right that in some cases we will report a stale leader here. However, once the asynchronous fetch from the `NodeCache` completes, the listener should be notified about the new leader. What happens in the meantime is that the listener will try to connect to the old leader which should either be gone or reject all connection attempts since he is no longer the leader.

The problem `notifyLossOfLeader` tried to solve is that a listener thinks that a stale leader is still the leader and, thus, continues working for it w/o questioning it (e.g. check with the leader) until the connection to the leader times out. With the `notifyLossOfLeader` change, once the retrieval service loses connection to ZooKeeper, it will tell the listener that the current leader is no longer valid. This will tell the listener to stop working for this leader (e.g. cancelling all tasks, disconnecting from it, etc.). If the listener should shortly after be told that the old leader is still the leader because of stale information, then it will first try to connect to the leader which will fail (assuming that the old leader is indeed no longer the leader) before it starts doing work for the leader.